### PR TITLE
[Bugfix][DeepSeek-V4] Fix Spec V2 Draft Input ID Dtype for DP Collectives

### PR DESCRIPTION
**Pull request by DaZhUUU**
## Motivation

This PR fixes a DeepSeek-V4 hang issue with DP attention, EAGLE speculative decoding, and overlap scheduling.

In the overlap spec-v2 draft-extend path, `batch_result.next_token_ids` can be `torch.int32` and is assigned to `batch.input_ids`. DeepSeek-V4 NextN can later consume `batch.input_ids` in DP collectives. If ranks enter the same collective with different input id dtypes, NCCL can hang because collective dtype/count/op must match across ranks.

The non-overlap speculative path already converts next token ids to `torch.int64`. This PR applies the same normalization to the overlap spec-v2 path.

## Modifications

Normalize `predict` to `torch.int64` in `EagleDraftInputV2Mixin.prepare_for_extend_to_fill_draft_kvcache()`:

```python
# Keep draft input ids as int64 to match the non-overlap spec path and
# avoid dtype mismatches in downstream DP collectives.
batch.input_ids = predict.to(torch.int64)
```

I also considered casting `input_ids` only near the DeepSeek-V4 NextN DP gather call. The current approach is better because it fixes the dtype at the source, keeps `batch.input_ids` consistent for all downstream consumers, and matches the existing non-overlap speculative path. It is also smaller and less model-specific than adding a DeepSeek-V4-only cast near one gather site.

## Accuracy Tests

This change only normalizes token id dtype before downstream DP collectives. It does not change logits computation, sampling logic, attention kernels, or model weights.

Validated with DeepSeek-V4-Flash using DP attention, EAGLE speculative decoding, and overlap scheduling.

Service command:

```bash
export SGLANG_ENABLE_THINKING=1
export SGLANG_DSV4_FP4_EXPERTS=1
export GLOO_SOCKET_IFNAME=<network-interface>
export NCCL_MIN_NCHANNELS=24
export NCCL_IB_QPS_PER_CONNECTION=8
export NCCL_DEBUG=INFO
export NCCL_DEBUG_SUBSYS=COLL

python3 -m sglang.launch_server \
  --trust-remote-code \
  --model-path <DeepSeek-V4-Flash-model-path> \
  --tp 8 \
  --dp-size 8 \
  --enable-dp-attention \
  --cuda-graph-max-bs 16 \
  --max-running-requests 128 \
  --enable-metrics \
  --host 0.0.0.0 \
  --port 8080 \
  --mem-fraction-static 0.85 \
  --moe-runner-backend marlin \
  --tool-call-parser deepseekv4 \
  --reasoning-parser deepseek-v4 \
  --speculative-algo EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --chunked-prefill-size 4096
```

The same configuration was also tested with `--disable-cuda-graph`.

Smoke test request:

```bash
curl -sS http://127.0.0.1:8080/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "DeepSeek-V4-Flash",
    "messages": [{"role": "user", "content": "Explain why distributed collectives require matching dtypes across ranks."}],
    "temperature": 0,
    "max_tokens": 256
  }'
```

Validation results:

```text
Default cuda graph:
- Service became ready successfully.
- 10 sequential curl requests completed successfully.
- Each response had non-empty content.
- No watchdog timeout, traceback, runtime error, or service hang was observed.
- NCCL logs did not show the previous problematic `AllReduce count=4 datatype=2` pattern.

No cuda graph:
- Service became ready successfully.
- 10 sequential curl requests completed successfully.
- 32 concurrent curl requests completed successfully: TOTAL=32 OK=32 FAIL=0.
- No watchdog timeout, traceback, runtime error, or service hang was observed.
- NCCL logs did not show the previous problematic `AllReduce count=4 datatype=2` pattern.
```

NCCL dtype check for `AllReduce count=4`:

```text
Default cuda graph:
   9111 datatype_4
    113 datatype_9
No datatype_2 examples found.

No cuda graph:
  18515 datatype_4
    259 datatype_9
No datatype_2 examples found.
```

## Speed Tests and Profiling

This PR is not intended to improve speed. The added cast is on draft token ids before DP collective usage, not on attention/MoE/kernel hot paths.

Expected overhead is small:

- If `predict` is already `torch.int64`, `Tensor.to(torch.int64)` should return the original tensor without a real copy.
- If `predict` is `torch.int32`, the cast is applied to a small token-id tensor, not to hidden states, logits, KV cache, or model weights.

Runtime smoke results did not show functional regression:

```text
Default cuda graph: 10 sequential requests completed.
No cuda graph: 10 sequential requests completed, and 32 concurrent requests completed with TOTAL=32 OK=32 FAIL=0.
```

No dedicated serving throughput benchmark was run for this one-line dtype normalization.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #27523252726](https://github.com/sgl-project/sglang/actions/runs/27523252726)
Latest PR Test (Extra): :x: [Run #27523252611](https://github.com/sgl-project/sglang/actions/runs/27523252611)

### CODE REVIEW

**Comment by DaZhUUU** [comment]
@hnyls2002 Hello, please review this PR. 

I found this bug and opened the issue https://github.com/sgl-project/sglang/issues/27938. 

After investigation, I traced it to an int32/int64 input id dtype mismatch in the EAGLE spec-v2 overlap path, which can hang DP collectives. I submitted a PR to normalize the draft input ids to torch.int64.

**Comment by DaZhUUU** [comment]
@hnyls2002 

I checked the latest upstream refactor. The affected spec v2 draft-extend logic has moved into `EagleDraftWorkerBase.prepare_for_draft_extend`, but the same issue still exists there: `predict` can be `int32` while `batch.input_ids` later reaches DeepSeekV4 DP gather, which requires a consistent integer dtype across ranks.

So even after the refactor, this path still needs the `int64` normalization before constructing `ForwardBatch`.

**Comment by hnyls2002** [comment]
/rerun-test test_eagle_dp_attention.py test_eagle_infer_beta_dp_attention.py test_eagle_worker_v2_topk1_fastpath.py test_spec_eagle.py
